### PR TITLE
Synchronize Laghos with MFEM master [okina]

### PR DESCRIPTION
**Pull request by camierjs**
This PR synchronizes `Laghos` with `MFEM` GPU support.
It allows to remove the subdirectories: `cuda`, `raja`, `hip` & `occa`.

---

- [x]  Sync serial code
- [x] `HIP` backend, serial & parallel
- [x] `valgrinded`

### CODE REVIEW

**Comment by tzanio** [comment]
Please also update `CHANGELOG` and `README.md`

**Comment by v-dobrev** [comment]
> It allows to remove the subdirectories: `cuda`, `raja`, `hip` & `occa`.

Does this mean that `hip` through mfem works fine here?

**Comment by camierjs** [comment]
> Does this mean that `hip` through mfem works fine here?

Yes, that's one of the purpose! I'll check again that the build is ok.

File: laghos.cpp [hunk 1]
```diff
@@ -585,38 +717,56 @@ int main(int argc, char *argv[])
...
+         const double e_norm = sqrt(norm);
+         MFEM_VERIFY(rs_levels==0 && rp_levels==0, "check: rs, rp");
+         MFEM_VERIFY(order_v==2, "check: order_v");
+         MFEM_VERIFY(order_e==1, "check: order_e");
+         MFEM_VERIFY(ode_solver_type==4, "check: ode_solver_type");
+         MFEM_VERIFY(t_final == 0.6, "check: t_final");
+         MFEM_VERIFY(cfl==0.5, "check: cfl");
+         MFEM_VERIFY(strncmp(mesh_file, "default", 7) == 0, "check: mesh_file");
+         MFEM_VERIFY(dim==2 || dim==3, "check: dimension");
+         Checks(dim, ti, e_norm, checks);
```
Comment (vladotomov) [inline laghos.cpp:763]: Are the `e_norm` numbers in `Checks` independent of the device?

File: laghos_assembly.hpp [hunk 2]
```diff
@@ -67,182 +67,111 @@ struct Tensors1D
...
+      auto d_x = x.Read();
+      auto d_y = y.Write();
+      MFEM_FORALL(i, N, d_y[i] = d_x[i] / d_diag[i];);
    }
-   virtual void SetOperator(const Operator &op) { }
-};
-
-// Performs partial assembly for the energy mass matrix on a single zone.
-// Used to perform local CG solves, thus avoiding unnecessary communication.
-class LocalMassPAOperator : public Operator
```
Comment (vladotomov) [inline laghos_assembly.hpp:222]: Is this replaced by something or just removed as a method?

File: laghos_solver.cpp [hunk 3]
```diff
@@ -78,130 +80,293 @@ void VisualizeField(socketstream &sock, const char *vishost, int visport,
...
    }
-   quad_data.h0 /= (double) H1FESpace.GetOrder(0);
+   qdata.h0 /= (double) H1.GetOrder(0);
 
    if (p_assembly)
    {
       // Setup the preconditioner of the velocity mass operator.
       Vector d;
-      (dim == 2) ? VMassPA.ComputeDiagonal2D(d) : VMassPA.ComputeDiagonal3D(d);
+      (dim == 2) ? VMassPA->ComputeDiagonal2D(d):VMassPA->ComputeDiagonal3D(d);
```
Comment (vladotomov) [inline laghos_solver.cpp]: I'll try replacing these with Andrew's PA Jacobi preconditioner tomorrow.

File: laghos.cpp (hunk 1 above)
Comment (camierjs) [reply laghos.cpp:763]: Yes, I'm trying to have all the results independent of the PA/FA/Device options.

Comment (tzanio) [reply laghos.cpp:763]: Maybe I misunderstand, but I will not expect identicall results between PA and FA (e.g. both ob CPU); or between PA on CPU and PA on GPU...

File: laghos_assembly.hpp (hunk 2 above)
Comment (camierjs) [reply laghos_assembly.hpp:222]: It's replaced for now by one global solve for the energy.
We can bring a local one if needed.

File: laghos.cpp (hunk 1 above)
Comment (camierjs) [reply laghos.cpp:763]: At least for the non-regression tests and the one in the `README`, that is useful though.

**Comment by vladotomov** [comment]
Now it's easy to add the diagonal Jacobi for the energy CG too, and it will probably improve the convergence. But it will change the autotest numbers. I'm fine either way.

**Review by vladotomov — APPROVED** [review]

File: laghos_assembly.hpp [hunk 4]
```diff
@@ -18,6 +18,8 @@
 #define MFEM_LAGHOS_ASSEMBLY
 
 #include "mfem.hpp"
+#include "general/forall.hpp"
+#include "linalg/dtensor.hpp"
```
Comment (v-dobrev) [inline laghos_assembly.hpp:22]: These paths (here and in a few other places) do not work when using an installed version of mfem.

One option is to add the flag `-I$(MFEM_DIR)/include/mfem` to the build options, if that directory exists.

As a longer term solution we probably need to add a new header `mfem-X.hpp` where `X` is something like `kernels`, `forall`, etc.

Comment (v-dobrev) [reply laghos_assembly.hpp:22]: I fixed this issue (for now) in df2dbed.

Comment (camierjs) [reply laghos_assembly.hpp:22]: Ok, thanks @v-dobrev, I sync the `serial` dir too with your changes.

Comment (v-dobrev) [reply laghos_assembly.hpp:22]: Ah, OK. I forgot about that directory. 😃

**Comment by camierjs** [comment]
The tests are now are `ok` with the `hip` backend up to order `2`.
Higher orders present the same behavior than the `CEED` benchmarks, even with the latest `rocm 3.1` stack.
The compiler is still really slow and gives asynchronous errors hard to debug.

**Review by v-dobrev — APPROVED** [review]

**Comment by v-dobrev** [comment]
I some warnings with `-pedantic -O3 -g -Wall`:
```ascii
./laghos_assembly.hpp:110:27: warning: private field 'ir' is not used [-Wunused-private-field]
   const IntegrationRule &ir, &ir1D;
                          ^
./laghos_assembly.hpp:129:26: warning: private field 'qdata' is not used [-Wunused-private-field]
   const QuadratureData &qdata;
                         ^
./laghos_assembly.hpp:130:24: warning: private field 'fes' is not used [-Wunused-private-field]
   FiniteElementSpace &fes;
                       ^
./laghos_assembly.hpp:135:15: warning: private field 'T1D' is not used [-Wunused-private-field]
   Tensors1D &T1D;
              ^
```

**Comment by camierjs** [comment]
> I some warnings with `-pedantic -O3 -g -Wall`:
> 
> ```
> ./laghos_assembly.hpp:110:27: warning: private field 'ir' is not used [-Wunused-private-field]
>    const IntegrationRule &ir, &ir1D;
>                           ^
> ./laghos_assembly.hpp:129:26: warning: private field 'qdata' is not used [-Wunused-private-field]
>    const QuadratureData &qdata;
>                          ^
> ./laghos_assembly.hpp:130:24: warning: private field 'fes' is not used [-Wunused-private-field]
>    FiniteElementSpace &fes;
>                        ^
> ./laghos_assembly.hpp:135:15: warning: private field 'T1D' is not used [-Wunused-private-field]
>    Tensors1D &T1D;
>               ^
> ```
Ok, good catch, let me take a look!

**Comment by camierjs** [comment]
Warning-free, which allowed to remove unused `Tensor1D`.

**Review by tzanio — APPROVED** [review]
It is so great to see Laghos in sync with the MFEM device support 🚀

I know this took a lot of hard work over a long period of time. Thank you Jean-Sylvain!

**Pull request merged**